File: NodeSerial.py
# NodeSerial.py - LASG/PBAI
# Created by Kevin Lam
# Updated by Adam Francey March 16 2018
# Handles serial connection and communication between Raspberry Pi and Node

import serial
from serial import SerialException
import time

class NodeSerial():

    def __init__(self, num_nodes):

        # num_nodes: number of nodes expected to be attached
        self.num_nodes = num_nodes

        #Initialized Serial Comm Locations, To Be Rearranged Later
        self.node_addresses = ['']*num_nodes

        #Teensy ID's in Physical Locations
        self.teensy_ids = [0]*num_nodes
        self.teensy_ids_bytes = [b'\x00\x00\x00']*num_nodes

        # list of serial ports
        self.serial_list = [None]*num_nodes

        # SOM (Start Of Message)
        # EOM (End Of Message)
        # this means that bytes 0xfe and 0xff
        # are OFF LIMITS for any other purpose!
        self.SOM = b'\xff\xff' # beginning of every message
        self.EOM = b'\xfe\xfe' # ending of every message
        self.padding_byte = b'\xfd'

        # fixed message length
        self.message_length = 25

        # password sent by node
        self.password = b'\xff\xff\x04\x00\x00\x01\xfe\xfe'

        # NOTE on bytes objects
        # example
        # byte_sequence = b'\xff\xfe\xfc'
        # byte_sequence[i] = integer (ie byte_sequence[0] = 255)
        # byte_sequence[i:i+1] = bytes object (ie byte_sequence[0:1] = b'\xff')
        
        self.baud_rate = 57600

        # register all port before initializing
        self.registerSerialPorts()

    def registerSerialPorts(self):
        
        # searches serial ports to find attached nodes
        # constructs self.node_addresses and self.serial_list

        prefix = '/dev/ttyACM' # linux serial port prefix, append integer for address
        #prefix = 'COM' # windows serial port prefix, append integer for address
        port_max = 100 # maximum number of ports to check
        port_timeout_limit = 2 # wait time in seconds to receive password from Teensy
        
        print("Beginning serial port registration")

        # find all nodes
        port_number = 0
        last_address_found = ''
        last_port_found = None
        num_registered_nodes = 0
        for node in range(self.num_nodes):
            print("Looking for node " + str(node) + "/" + str(self.num_nodes))

            # try to open serial ports until a node is found
            teensy_found = False
            while teensy_found == False and port_number < port_max:

                try:
                    # try to open port
                    # the try block will exit after this line if a serial device is not found
                    ser = serial.Serial(prefix + str(port_number), self.baud_rate)

                    print("Serial device found on " + prefix + str(port_number) + ": ", end = '')
                    
                    # if this is a node, it will be continuously sending a password until it receives a password back
                    # so we check the serial port to see if we get the password sequence
                    password_received = False
                    timeout = False
                    password_index = 0
                    start_time = time.time()
                    while password_received == False and timeout == False:
                        if ser.inWaiting() > 0:
                            # check if it matches the current byte in password sequence
                            incoming_byte = ser.read()
                            if incoming_byte == self.password[password_index:password_index+1]:
                                # if so, we increase index to check for next byte on next loop
                                password_index += 1
                            else:
                                # else, start looking from beginning
                                password_index = 0

                        if password_index == len(self.password):
                            print("Teensy detected")
                            # we have received the password!! This is a node!
                            password_received = True
                            teensy_found = True
                            last_address_found = prefix + str(port_number)
                            last_port_found = ser

                        if time.time() - start_time > port_timeout_limit:
                            # we have waited long enough, it is (probably) not a node
                            print("Not a Teensy (timed out)")
                            timeout = True
                            ser.close()
                                   
                    
                except SerialException:
                    # there is no serial device on this port
                    # do nothing
                    pass

                port_number += 1

            if teensy_found == True:

                # if we found a device, register it with system
                self.node_addresses[node] = last_address_found
                self.serial_list[node] = last_port_found

                # send password back to Teensy
                self.serial_list[node].write(self.password)

                # ditch extra password bytes in buffer, we don't need them anymore
                self.serial_list[node].reset_input_buffer()

                # wait for Teensy to send back Teensy ID
                time.sleep(1)

                if (self.serial_list[node].inWaiting()>2):
                    tid1 = self.serial_list[node].read()
                    tid2 = self.serial_list[node].read()
                    tid3 = self.serial_list[node].read()
                    self.teensy_ids_bytes[node] = tid1+tid2+tid3
                    self.teensy_ids[node] = ((tid1[0] << 16) | (tid2[0] << 8)) | tid3[0]

                    

                num_registered_nodes =+ 1
            else:
                # otherwise, we have hit port max
                print("Port maximum (" + str(port_max) + ") reached: " + str(port_number>=port_max))

        # print out info for registered nodes
        print(str(num_registered_nodes) + " out of " + str(self.num_nodes) + " nodes found ")
        for node in range(num_registered_nodes):
            print("Node: " + str(node) + " - Address: " + self.node_addresses[node] + " - ID: " + str(self.teensy_ids[node]))
        

    def waitUntilSerialPortsOpen(self):
        # DEPRECATED ---------------------------------------------------------------
        # only continue once we are able to open all serial ports
        print("Attempting to open serial ports...")
        while True:
            try:
                for i in range(len(self.serial_list)):
                    self.serial_list[i] = serial.Serial(self.node_addresses[i],self.baud_rate)
                    self.serial_list[i].flush()
                    self.serial_list[i].flushInput()
                break # if we are able to open all, break out of loop

            # catch the exception, wait one second before trying again
            except SerialException:
                time.sleep(1)
        print("Serial Ports Opened Sucessfully")

    def rearrangeSerialPorts(self):
        # NEEDS TO BE UPDATED -------------------------------------------------------------
        for i in range(len(self.serial_list)):
            sendMessage(INSTRUCT_CODE_GET_TEENSY_ID)
            code, tid = checkIncomingMessageFromNode(i)

    def checkIncomingMessageFromNode(self, node_number):

        # the raw bytes coming from serial port
        raw_bytes = b''

        #Serial port for Target Node
        ser = self.serial_list[node_number]

        #Check if port has minimum message length in buffer
        if ser.inWaiting() > 6: # at lseat enough bytes for SOM, ID, length, code

            #Check for Start of Message
            for i in range(len(self.SOM)):
                current_SOM = ser.read()
                raw_bytes += current_SOM
                if current_SOM != self.SOM[i:i+1]:
                    # Fail: SOM byte missing
                    # CheckIncomingMessageFromNode fails, stop reading serial port
                    print("SOM Not Found")
                    print(current_SOM)
                    return "error", "SOM missing", None, None

            #Teensy IDs, TODO: Use these for validation
            t1 = ser.read()
            t2 = ser.read()
            t3 = ser.read()
            raw_bytes = raw_bytes + t1 + t2 + t3
            received_tid = ((t1[0] << 16) | (t2[0] << 8)) | t3[0]
            
            # Read in Length and Code
            #data_length = int.from_bytes(ser.read(), byteorder = 'big')
            data_length = ser.read()
            raw_bytes += data_length
            data_length = data_length[0] # taking first element of one-element bytes object changes the byte to int
            message_code = ser.read()
            raw_bytes += message_code

            # read data
            incoming_data = []

            if ser.inWaiting() > data_length + 1: # at least enough bytes for data and EOM

                # read in data bytes (could be no bytes)
                for i in range(data_length):
                    data_byte = ser.read()
                    incoming_data.append(data_byte)
                    raw_bytes += data_byte

                #Check for End of Message
                for i in range(len(self.EOM)):
                    current_EOM = ser.read()
                    raw_bytes += current_EOM
                    if current_EOM != self.EOM[i:i+1]:
                        # Fail: EOM byte missing
                        # stop reading serial port, all read bytes are unused (aka discarded)
                        return "error", "EOM missing", None, None
                # if we get here, we have found EOM and therefore whole message
                

                # Success: return message
                # Returns identifier byte for message code and relevant data list
                print("Received message from node " + str(node_number) + " - Code: " + str(message_code) + ", Data: " + str(incoming_data))
                return message_code, incoming_data, received_tid, raw_bytes
            else:
                return "error", "not enough data bytes", None, None


        else:

            return "error", "not enough bytes before data", None, None

    def sendMessage(self, outgoing_message_code, outgoing_data, node_number):
        # input:
        # outgoing_message_code: bytes object of length 1
        # outgoing_data: list of ints (can be empty), ASSUMES EACH ITEM CAN BE TRANSFORMED INTO ONE SINGLE BYTE EACH
        # node_number: int

        data_length = len(outgoing_data) # outgoing data bytes

        # select serial port
        ser = self.serial_list[node_number]

        # the bytes object the we will construct and send to the serial port
        message = (self.SOM # 2 bytes
                   + self.teensy_ids_bytes[node_number] # 3 bytes
                   + bytes([data_length]) # 1 byte
                   + outgoing_message_code # 1 byte
                   + bytes(outgoing_data) # len(outgoing_data) bytes
                   + self.EOM) # 2 bytes

        ser.write(message)

        print("Sending message to node " + str(node_number) + " - Code: " + str(outgoing_message_code) + ", Data: " + str(outgoing_data))

    def getNodeNumberFromId(self, teensy_id):
        for i in range(len(self.teensy_ids)):
            if teensy_id == self.teensy_ids[i]:
                return i

    def close(self):
        for i in range(self.num_nodes):
            print ("Stopping port: " + str(self.node_addresses[i]))
            self.serial_list[i].close()
        

if __name__ == '__main__':

    num_nodes = 1

    # test message
    TEST_LED_PIN_AND_RESPOND = b'\x00'
    num_blinks = 10

    S = NodeSerial(num_nodes)

    S.sendMessage(TEST_LED_PIN_AND_RESPOND, [num_blinks], 0)

    time.sleep(1) # wait a second for response
                  # this is not needed on windows
                  # is needed on Pi

    code, data, tid, raw_bytes = S.checkIncomingMessageFromNode(0)
    print("Message received from node 0 - Code: " + str(code) + ", Data: " + str(data))
    print("Raw bytes of message: " + str(raw_bytes))
    

    while True:
        try:

            # just some code to fill in this gap (this will be where main program code goes
            # for some reason if it is empty the KeyboardInterrupt doesn't get caught below
            for i in range(100):
                filler = 0
        except KeyboardInterrupt:
            print("Closing Main Program and Serial Ports")
            S.close()

            print("Completed, Goodbye!")
            break

File: LuddyNode/NodeSerial.cpp
/*
* NodeSerial.cpp - Library for Communication Protocol between Raspberry Pi and Node
* created By Adam Francey March 13, 2018
* Modified from usb_serial_comm.cpp created By Adam Francey, Kevin Lam, July 5, 2017
* Released for Luddy Hall 2018
* Philip Beesley Architect Inc. / Living Architecture Systems Group
*
* Message Format (see related doc [WIP] for details)
* where <value> is the value represented as a byte:
* <SOM1><SOM2><TeensyId1><TeensyId2><TeensyId3><total number of bytes in message><code><data1>...<dataN><EOM1><EOM2>
*/

#include <Arduino.h>
#include <stdint.h>
#include "NodeSerial.h"

NodeSerial::NodeSerial(){
  Serial.begin(57600);
}

NodeSerial::NodeSerial(int baud_rate){
  Serial.begin(baud_rate);
}

NodeSerial::~NodeSerial(){}

void NodeSerial::SendMessage(uint8_t code){
  // Send SOM
  for (int i = 0; i < NUM_SOM; i++){
    Serial.write(SOM[i]);
  }

  // Send TeensyID
  for (int i = 0; i < NUM_ID; i++){
    Serial.write(ID[i]);
  }

  // Send message length
  //                 SOM       ID       length   code  EOM
  uint8_t message_length = NUM_SOM + NUM_ID + 1      + 1   + NUM_EOM;
  Serial.write(message_length);

  // Send code
  Serial.write(code);

  // Send EOM
  for (int i = 0; i < NUM_EOM; i++){
    Serial.write(EOM[i]);
  }

}

void NodeSerial::Register(uint8_t my_id_bytes[]){

  while (initialized == false){

    // send password
    for (int i = 0; i < password_length; i++){
      Serial.write(password[i]);
    }

    // delay okay to use here
    // it is fine to use before the sculpture initializes
    delay(100);

    // check serial port for password sequence
    // this means the Raspberry Pi has recognized the Teensy
    bool eligible = true;
    bool password_received = false;
    if (Serial.available() > password_length - 1){
      for (int i = 0; i < password_length; i++){
        if (Serial.read() != password[i]){
          eligible = false;
        }
      }
    
    } else {
      eligible = false;
    }
  
    if (eligible){
      password_received = true;
    }
    initialized = password_received;
  }

  // wait for Raspberry Pi to flush its buffer
  delay(500);

  //send Teensy ID bytes
  Serial.write(my_id_bytes[0]);
  Serial.write(my_id_bytes[1]);
  Serial.write(my_id_bytes[2]);

  // let NodeSerial know what the id is
  ID[0] = my_id_bytes[0];
  ID[1] = my_id_bytes[1];
  ID[2] = my_id_bytes[2];
}


void NodeSerial::SendMessage(uint8_t code, uint8_t data[], uint8_t data_length){

    // Send SOM
    for (int i = 0; i < NUM_SOM; i++){
        Serial.write(SOM[i]);
    }

    // Send TeensyID
    for (int i = 0; i < NUM_ID; i++){
        Serial.write(ID[i]);
    }

    // Send data length
    Serial.write(data_length);

    // Send code
    Serial.write(code);
    
    // Send data
    for (uint8_t i = 0; i < data_length; i++){
        Serial.write(data[i]);
    }

    // Send EOM
    for (int i = 0; i < NUM_EOM; i++){
        Serial.write(EOM[i]);
    }

}


// CheckMessage
// Reads the serial port for an incoming sequence of bytes
// if sequence follows requirements of a message defined in docs, returns 1 and
// updates the following NodeSerial members:
// last_data_length, last_code_received, last_data_received, message_waiting
// if requirements fail, returns 0
bool NodeSerial::CheckMessage(){

  message_waiting_ = 0;

  //if (Serial.available()>=MESSAGE_LENGTH){
  if (Serial.available()>NUM_SOM + 4){ // at least 7 bytes for SOM, ID, length, code

    // Teensy ID
    uint8_t t1;
    uint8_t t2;
    uint8_t t3;

    // modifiers
    uint8_t data_length;
    uint8_t code;

    // check for SOM (bytes 0-1)
    // we are expecting SOM bytes in sequence (currently: {0xff,0xff})
    for (int s = 0; s < NUM_SOM; s++){
      if (Serial.read() != SOM[s]){
        // Fail: SOM byte missing
        // So CheckMessage fails (no message found) and we stop reading serial port
        return 0;
      }
    }
    // if we make it here, we have found all SOM bytes

    // Teensy ID (bytes 2-4)
    t1 = Serial.read();
    t2 = Serial.read();
    t3 = Serial.read();

    // data_length (byte 5)
    data_length = Serial.read();

    // command/code (byte 6)
    code = Serial.read();

    // SOM (2 bytes), Teensy ID (3 bytes), message length (1 byte), code (1 byte) should take up 7 bytes
    // we are now expecting num_bytes_to_receive more bytes (including EOM)
    uint8_t data[data_length];

    if (Serial.available()>data_length+1){ // at least enough bytes for data and EOM

      for (uint8_t i = 0; i < data_length; i++){
        // read in data bytes
        data[i] = Serial.read();
      }
  
      // check for EOM
      for (int e = 0; e < NUM_EOM; e++){
        if (Serial.read() != EOM[e]){
          // Fail: EOM byte missing
          // something went wrong: num_bytes_in_message tells us to expect EOM here but we didn't find it
          // CheckMessage fails, stop reading serial port
          // bytes read up until this point will be discarded
          return 0;
        }
      }
      // If we make it here, we have found all EOM bytes
      
      // Success: Full message found
      // update class members
      // note: this will be overwritten on the next message successfully recieved
      last_data_length_ = data_length;
      last_code_received_ = code;
      for (int i = 0; i < data_length; i++){
        last_data_received_[i] = data[i];
      }
      message_waiting_ = 1;
      return 1;
    } else {
      //Fail: not enough bytes
      return 0;
    }



  } else {
    // Fail: not enough bytes
    return 0;
  }
}
